hdf5_serialiser/__init__.py

```diff
@@ -1,14 +1,12 @@
 __all__ = ["serialisable"]
 
-import dataclasses
+from dataclasses import is_dataclass, fields
 
 from pathlib import Path
-from typing import IO, Any, Type, TypeVar
-import typing
+from typing import IO, Any, Protocol, Type, Union, get_args, get_origin, runtime_checkable
```

Comment:
I'm not a big fan of importing everything into global namespaces. While Things like Any, Union, etc are pretty obvious to be coming from `typing`, others, like `get_args` aren't. Similar argument for BaseModel. If I was defining several pydantic models here, I'd certainly use `from pydantic import BaseModel`. However, here, I'm only using this for a single check and I find it much clearer to use `pydantic.BaseModel`. It's very obvious when you read the code what it is. Bikeshedding, so feel free to leave as it is. In a nutshell: I think explicit is better than implicit.

Reply:
Idk about this - importing from the same module twice feels odd to me - either style works in terms of doing typing.x everywhere, or importing everything individually, but I don't think you can have it both ways. Otherwise typing.Union and Union are equivalent. 

> While Things like Any, Union, etc are pretty obvious to be coming from typing, others, like get_args aren't.

This is part I don't buy - to me "get_args" is just as obvious as anything else, but maybe I've been immersed in types too long haha

Reply:
It's kind of like if I said, most numpy functions are obvious to me, so I'll import directly:
```py
from numpy import array, sum, multiply
```

...but einsum is a strange function, so I will import in a more explicit way:

```py
import numpy
numpy.einsum(...)
```

Reply:
Maybe the argument wasn't great. I rarely use `from ... import ...` because I think it's saving extra few characters isn't worth it. But I do often use it for very limited number of imports, one of them is typing (although not so much anymore since you usually don't need `Union`, `Dict`, `List` -- these are replaced by `|`, `dict`, `list` now. I can certainly imagine that I'd write my own function which would be called `get_args`, but I really doubt I'd put it in a file called `typing` if you see what I mean ;)

Reply:
Another example where I often use `from .. import ...` is `from datetime import datetime`. Also if the module's name is long or has many dots, I tend to use `import very.long.module.name as shorter`. The bottom line is: It's impossible to remember where the functions/names are coming and... you don't need to remember them if you use `module.name`.

Reply:
This seems to support it: https://docs.python-guide.org/writing/structure/#modules
I don't know who wrote it, so it might need to be read with a pinch of salt though.